### DISCUSSION TITLE
3D data in INSPIRE geology model

### DISCUSSION THREAD

**Discussion by KnorrAn** [INSPIRE-MIF/helpdesk#149] [General]
Dear all,

I have a question according 3D data in the INSPIRE data model geology.
We are tasked to provide 3D data from a geological subsurface model as interoperable INSPIRE services, but we are not sure if the data specification allows this at all. Our source data are in XYZ coordinates (see screenshots, left: in 3D viewer, right: in QGIS 2D).
From the TG and the IR Reqiurements Article 12 I can infer that spatial features are basically limited to the Simple Feature Access of the OGC, i.e. they may have a maximum of 2.5 D. Does anyone in the community have any experience with implementing 3D data in INSPIRE or know who I can contact? That would be really helpful. 

Looking forward to a lively exchange!

Thanks!
Antje

![grafik](https://user-images.githubusercontent.com/124601863/227963790-d29d52b7-63b9-4288-adff-3b4e601c1a7f.png)

**Comment by reinij** [comment]
Deat Antje,

I'm not an expert in geology data model, but for example with buildings, INSPIRE has 3D data model https://inspire.ec.europa.eu/Themes/126/2892 and https://inspire.ec.europa.eu/data-model/approved/r4618-ir/html/index.htm?goto=2:3:2:3:7916

It seems to be that this is not the case with geology. As a wild guess I would propose to initiate a process for "Good Practices" (https://inspire.ec.europa.eu/portfolio/good-practice-library). This process is more flexible than TG changes, but could be a "quick" solution and also endorsed by EC if there are implementations available.

**Comment by sMorrone** [comment]
Dear @KnorrAn,
In the context of the GO-PEG project, ISPRA - Geological Survey of Italy has developed an efficient approach to manage and deliver interoperable subsurface geological data using an INSPIRE-extended data model and [GeoPackage INSPIRE alternative encoding](https://github.com/INSPIRE-MIF/gp-geopackage-encodings/blob/main/spec/GeoPackage_Good_Practice_initiation_fiche.md). 
This approach fulfills the institutional mandates of the Geological Survey of Italy and has been the starting point for the ongoing design and implementation of the “**Geological 3D subsurface models database**” related to the National Geological Mapping Programme.  
Geologic Units and Faults are the addressed spatial objects. 
In the sake of the ease of use and interoperability, it was decided to operate a simplification for the geometry representation of the Geologic Units (in 3D model described by volumes), i.e., not directly deal with ‘real’ volumes but provide subsurface geometry information of the Geologic Units through related “full” surfaces of Geologic Unit Boundaries.
This means that a discretization is operated of the Geologic Unit Boundary surfaces so that they are collections of points (i.e., they are “full” surfaces). Each point of these surfaces is a spatial object characterized by **x, y, z values plus a thickness (t)** value 
![image](https://user-images.githubusercontent.com/13329248/228460542-fc35b369-cc23-4253-9035-afeb9f1ae755.png)
You can find more details and download a sample dataset from the [GeoPackage INSPIRE Good Practice implementation examples page](https://github.com/INSPIRE-MIF/gp-geopackage-encodings/blob/main/examples/overview.md) (last row of the table). You can also find a [related power point presentation](https://inspire.ec.europa.eu/sites/default/files/gpkg_inspire_subsurface.pdf) at [the INSPIRE Good Practice – GeoPackage and implementation practice Webinar page](https://inspire.ec.europa.eu/events/inspire-good-practice-geopackage-and-implementation-practice-webinar) and post any additional questions /comments you may have in the INSPIRE-MIF GitHub [repository for the GeoPackage encodings](https://github.com/INSPIRE-MIF/gp-geopackage-encodings)

**Comment by sMorrone** [comment]
Small helpful details below on the GeoPackage sample.
When opening in QGIS, if you select the "Po Basin subsurface" (see first screenshot below) the layers open with pre-set tables relations and layer styles and the related information is nicely displayed in the same form (see screenshot 2)
![image](https://user-images.githubusercontent.com/13329248/228467814-85de0e20-16bf-4809-9973-083465626d89.png)

![image](https://user-images.githubusercontent.com/13329248/228469619-493b60c4-422e-4b88-83fe-cc7b8474c269.png)

**Comment by HanWelmer** [comment]
My assumption, based on the Geology dataset specifications, is that 3D geometry should be possible. We at Geological Survey of the Netherlands - part of TNO - will try to use a 3D voxel grid for one of our models.

**Comment by KathiSchleidt** [comment]
Hi @HanWelmer 
have you investigated utilizing Coverage models for this data? I've just scanned the INSPIRE data models on GE, only use of Coverages I could find was under Hydrogeology, HydrogeologicalSurface on GroundWaterBody.
I'm not a geologist, so not quite sure which part of GE the data you're trying to encode, but assuming the core Geology Schema. At present, this is not foreseen, but to my view, a 3D coverage encoding of such data would make a great deal of sense, as such a model would allow for easy provision of such voxel information.
I'm happy to discuss, provide guidance on how this could be implemented.

**Comment by HanWelmer** [comment]
Hallo Kathi,
Thanks for the message.
So far we have created a 2D coverage (pixels) for RenewableAndWastePotentialCoverage in theme ER (see dataset with UUID 11a292bd-9022-4a9f-b1bf-02f5c89ad74a and with title "Totale potentie aardwarmte uit aquifers onder INSPIRE"). Any feedback from your side is always welcome.
I expect no problems repeating the trick for a 3D coverage (voxels) as and when it comes so far.

**Comment by KathiSchleidt** [comment]
Hi Han,
Exactly, same trick would apply. But, you'd first need to agree on an extended data model for this data, as at present there are no coverage models foreseen for core Geology. One would need to set up the necessary [INSPIRE Good Practice](https://inspire.ec.europa.eu/portfolio/good-practice-library).
Also a more general question on provision of coverage data - do you plan to only provide the coverage data via ATOM feed as you currently do? The ER dataset is coarse enough that the entire dataset for NL is digestible (632KB), but coverage files can rapidly become unmanageable. Have you considered utilizing a [Web Coverage Service (WCS)](https://inspire.ec.europa.eu/good-practice/ogc-compliant-inspire-coverage-data-and-service-implementation) for this?

**Comment by KnorrAn** [comment]
Hi all,

first of all: Thanks a lot for your ideas and suggestions!

@reinij: I actually find a process for "good practices" a bit "wild" at the moment. But thanks for pointing it out. Maybe something will come up in this direction later...

@sMorrone: Thanks, for the detailed presentation of your implementation. I saw your presentation at the webinar. Geopackage seems like a great way to publish complex data. As I understand, this is an additional encoding and GML is still mandatory, right? What I would be interested to know is if you have also published a view and download service of the 3D data as GML. If yes, would it be possible to provide a sample dataset of your implementation in Hale?
You write about an INSPIRE-extended data model. Where can I find or read something about it?

@HanWelmer and @KathiSchleidt: Thanks for your input! Unfortunately I have no experience with 3D voxels and WCS. Our source data are vector based TINs of the surfaces. Thematically, they are seismic reflection interfaces and faults. So definitely no hydrological data and, as I understand right, consequently WCS ist not possible yet.

**Comment by sMorrone** [comment]
Dear @KnorrAn,
The GeoPackage can be used as an additional or alternative encoding. For validation purposes, however, a GML (even a sample) must be derived from the GeoPackage and validated with the INSPIRE Validator. Usually, the export to INSPIRE GML is not too complex (as the GeoPackage structure is derived from the same UML model). In addition, if you use a GeoPackage structure for which there is already a GPKG to GML transformation project - e.g. Hale Studio project - this can be reused because the source (GPKG ) and target (INSPIRE XSD) structures would be the same ;) 
Regarding the services, you can find API collections at: https://geoserver.epsilon-italia.it/geoserver/ogc/features/collections.  If you want to visualize directly in QGIS, clearly use the URL to landing page: https://geoserver.epsilon-italia.it/geoserver/ogc/features.
Note that, while downloading the geopackage file from the [link](https://www.epsilon-italia.it/public/GO-PEG/Po_Basin_subsurface.gpkg) provided in the Good Practice repository allows you to also see the styles and the table relations (see [my previous reply](https://github.com/INSPIRE-MIF/helpdesk/issues/145#issuecomment-1488135092), the same cannot currently be done downloading data from the APIs. That's why in the APIs DB views are also shared (i.e. the **Po Plain - Geologic Units subsurface** and the **Po Plain Faults - Points** collections).
As for the data model extension, find it below (split in two diagrams for better readability - extension in green):


![go-depth-GeologicUnit](https://user-images.githubusercontent.com/13329248/229564940-4206ca1d-9671-4fa2-bfe9-b0b9a68f38c0.jpeg)
![go-depth-GeologicStructure](https://user-images.githubusercontent.com/13329248/229565079-46d4693a-9e9f-4268-b045-9e9ecd9d4dce.jpeg)
